internal/subscriber/subscriber.go

```diff
@@ -571,7 +571,14 @@ func (s *Subscriber) pull(req *PullRequest) {
 				continue
 			}
 			ackID := ackMessage.BuildAckID()
-			sm = append(sm, &metrov1.ReceivedMessage{AckId: ackID, Message: protoMsg, DeliveryAttempt: msg.CurrentRetryCount + 1})
+
+			// check if the subscription has any filter applied and check if the message satisfies it if any
+			if s.checkFilterCriteria(ctx, protoMsg) {
+				sm = append(sm, &metrov1.ReceivedMessage{AckId: ackID, Message: protoMsg, DeliveryAttempt: msg.CurrentRetryCount + 1})
+			} else {
+				// self acknowledging the message as it does not need to be delivered for this subscription
+				s.acknowledge(ackMessage.(*AckMessage).WithContext(ctx))
```

Comment:
ack to subscriber means that message was consumed. shoudn't this offset be commited directly to broker ?

Reply:
So to do that, we will have to handle the whole filtering in the consumer itself right? And if so, we will have to handle filtering logic for all the consumer implementations in future. It will lead to having same thing handled at multiple places. That's why I handled the filtering in subscriber, after the consumer has fetched the messages from the broker. 

Thoughts on this?

Reply:
Or we can handle the filtering in subscriber only, but before the offset management happen in subscriber and commit the skipped messages directly to consumer from there.